boot/zephyr/serial_adapter.c

```diff
@@ -117,19 +118,25 @@ boot_uart_fifo_callback(struct device *dev)
 	u8_t byte;
 	int rx;
 
-	while (uart_irq_update(uart_dev) &&
-	       uart_irq_rx_ready(uart_dev)) {
+	uart_irq_update(uart_dev);
 
+	if (!uart_irq_rx_ready(uart_dev)) {
+		return;
+	}
+
+	while (true) {
 		rx = uart_fifo_read(uart_dev, &byte, 1);
```

Comment:
So in case when interrupt is fired for each byte we can expect rare call to this which will return 0 bytes.